bits vs. bitcoins setting

peter-conalgo opened the pull request:
I started to use Breadwallet on my new phone and prefer it to all the other options. While I agree that _bits_ are 100% the right unit to be promoting to new users, as a "long-time" bitcoin user, I sometimes want to work in (fractional) bitcoins instead of bits.

This pull requests adds a new setting that allows the user to select between "bits" and "full bitcoins" for display of amounts (and entry). It does not offer "mBTC" as a choice. "bits" remains the default settings.

I also removed a hidden feature which implemented the same setting: by swiping the nav bar header when choosing local currency. Instead, now if you swipe in the navbar of the fiat currency chooser, it will now alternate between showing the value of one dollar, vs the value of one bitcoin ($1=15,000 bits and 1,000,000 bits = $666 USD). The old setting (if it was set to BTC instead of bits or mBTC) will be preserved and copied over to the new setting on first run.

I took the liberty of adding a few new events to the tracking system to help understand the usage of this setting. Whatever backend is consuming those events will need to be updated with these new events: `settings:show_denom_selector`, `settings:pick_btc` and `settings:pick_bits`.

TODO: There are a few new strings to be translated. Since their text may change during the approval process, I haven't isolated them yet.

![screen shot 2016-07-14 at 4 30 10 pm](https://cloud.githubusercontent.com/assets/1482781/16854740/4ec85432-49e0-11e6-9ff2-95282cde2b37.png)

aviadoffer commented:
Finally. After all those conversations for something so simple.
Thanks for listening. once it's out I might switch back to bw.

ffxsam commented:
Agreed. This is a great addition, hopefully it gets merged!

voisine commented:
Very nice. It's good to have a PR like this available for those who want it. The code looks good as well.

Several people I've interacted with on this subject are big fans of mBTC though, so I'd suggest you keep it in this PR. Also, even if you don't like using the SETTINGS_MAX_DIGITS_KEY to store the denomination setting, I'd recommend you use it to minimize the diff. It'll be easier to avoid conflicts, keep it rebased, and prevent any issues going back and forth between this patch and the main branch.

abgit commented:
:+1:

jonasschnelli commented on BreadWallet/BRSettingsViewController.m:
```diff
@@ -545,13 +574,37 @@ - (void)showCurrencySelector
+    
+    // XXX prepare this for i18n once we've finalized the text
+    [options addObject:BITS @" = bits (recommended)"];
+    [options addObject:BTC @" = Bitcoin"];
```
Shouldn't it be `Bitcoins` (mind the **s**) to be consistent?

jonasschnelli commented:
Nice PR.
Concept ACK

peter-conalgo commented:
Great feedback! I've added "mB" support and switched back to using `SETTINGS_MAX_DIGITS_KEY` with contents of 2, 5, or 8. Also added the 's' to Bitcoins in the menu, as follows:

![screen shot 2016-07-15 at 9 18 36 am](https://cloud.githubusercontent.com/assets/1482781/16875447/7e67541c-4a6d-11e6-8302-fd590399ef8f.png)

dabura667 commented:
Nice! I'll point people to this branch when they need to use other units.

bwfan commented:
BIG user of bread wallet and have been waiting for this feature!

@voisine I'm lost. Are you going to merge or just keep as a fork?

ffxsam commented:
I'm going to nitpick a bit here:

"Bitcoin" should be lowercase "bitcoin." For one, there are already inconsistencies in the update, e.g. "one bitcoin = mB 1,000" but then later Bitcoins is capitalized.

Capitalized "Bitcoin" refers to the name of the technology. Lowercase "bitcoin" refers to the monetary unit.

aviadoffer commented:
Other wallets use BTC vs bits.

On Fri, Jul 15, 2016 at 1:43 PM, Sam <notifications@github.com> wrote:

> I'm going to nitpick a bit here:
>
> "Bitcoin" should be lowercase "bitcoin." For one, there are already
> inconsistencies in the update, e.g. "one bitcoin = mB 1,000" but then later
> Bitcoins is capitalized.
>
> Capitalized "Bitcoin" refers to the name of the technology. Lowercase
> "bitcoin" refers to the monetary unit.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/voisine/breadwallet/pull/387#issuecomment-233018871>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHw12ZoD5BFiatsxf7DKKwS8JFdiURWEks5qV8bBgaJpZM4JM0mQ>
> .
>

ffxsam commented:
>Other wallets use BTC vs bits.

BTC is an appropriate abbreviation for "bitcoins." Like USD = dollars.

peter-conalgo commented:
You're right that I'm inconsistent even on that one screen shot... I will change to all lowercase.

aviadoffer commented:
https://en.wikipedia.org/wiki/Bitcoin#Units


On Fri, Jul 15, 2016 at 1:45 PM, Sam <notifications@github.com> wrote:

> Other wallets use BTC vs bits.
>
> BTC is an appropriate abbreviation for "bitcoins." Like USD = dollars.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/voisine/breadwallet/pull/387#issuecomment-233019428>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHw12TEPAv6V9gvKSKub7keCo-ncg3rAks5qV8dHgaJpZM4JM0mQ>
> .
>

peter-conalgo commented:
Updated screenshot:
![screen shot 2016-07-15 at 1 47 57 pm](https://cloud.githubusercontent.com/assets/1482781/16883630/f0a12eb6-4a92-11e6-93e3-1f72b01e0645.png)

voisine commented:
> BTC is an appropriate abbreviation for "bitcoins." Like USD = dollars.

Although its common to use BTC, it can't be official because anything starting with BT denotes that it is issued by the government of Bhutan. Non-state commodities like silver and gold start with X, XAU and XAG... also XBC is already taken for some european bond market unit, so XBT is probably what will end up being the ISO recognized abbreviation.

bwfan commented:
@voisine Are you going to merge or just keep as a fork?

voisine commented:
@bwfan, no plans to merge for reasons previously explained at excruciating length, but it's a nice patchset. I'm happy to point anyone to it who's really dead set on not switching to bits.

bwfan commented:
Ok. :(

aviadoffer commented:
figures. waste of time.


On Fri, Jul 15, 2016 at 2:15 PM, bwfan <notifications@github.com> wrote:

> Ok. :(
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/voisine/breadwallet/pull/387#issuecomment-233027361>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHw12SqZFN4OjCpk_u2MfwYpLWm51oYoks5qV85PgaJpZM4JM0mQ>
> .
>

voisine commented:
> figures. waste of time.

Definitely not a waste of time. Many long time bitcoin users want extra features for themselves that don't make sense for a streamlined, UX focused mass market product. Now they can easily merge this PR.

https://twitter.com/JoeSalowitz/status/747869196848029696

https://hackernoon.com/the-feature-fallacy-why-more-features-dont-make-a-better-product-9fd125e86440

aviadoffer commented:
That's your personal opinion. If you set it up for vote I bet you will get
a different view.
Giving the option doesn't take a way from the streamlined approach. Those
who don't know it will not use it anyway and default to bits just like your
vision.
In any case the argument was chewed too long too many times. No point to
rehash it.


On Fri, Jul 15, 2016 at 2:23 PM, Aaron Voisine <notifications@github.com>
wrote:

> figures. waste of time.
>
> Definitely not a waste of time. Many long time bitcoin users want extra
> features for themselves that don't make sense for a streamlined, UX focused
> mass market product. Now they can easily merge this PR.
>
> https://twitter.com/JoeSalowitz/status/747869196848029696
>
>
> https://hackernoon.com/the-feature-fallacy-why-more-features-dont-make-a-better-product-9fd125e86440
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/voisine/breadwallet/pull/387#issuecomment-233029524>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHw12XenA22QgmNbaiEDPkAzD5eBRJo0ks5qV9A6gaJpZM4JM0mQ>
> .
>

bwfan commented:
That's very extremist. I hope it forks and goes to the store with another name, see the Node vs Joynet mess.

voisine commented:
@aviadoffer @bwfan there's already several forks that are in the app store. You can probably find them in the list here: https://github.com/voisine/breadwallet/network/members

Now any of them can merge this PR, or you can make your own fork and do it with about two clicks. It's even better than voting. You don't need to convince anyone to get what you want.

aviadoffer commented:
Thanks Aaron. Always armed with a smart answer.(that a compliment)

I call for any one of the fork leaders with an app on the store to adopt
this lovely small change and identify the app name :)


On Fri, Jul 15, 2016 at 2:36 PM, Aaron Voisine <notifications@github.com>
wrote:

> @aviadoffer <https://github.com/aviadoffer> @bwfan
> <https://github.com/bwfan> there's already several forks that are in the
> app store. You can probably find them in the list here:
> https://github.com/voisine/breadwallet/network/members
>
> Now any of them can merge this PR, or you can make your own fork and do it
> with about two clicks. It's even better than voting. You don't need to
> convince anyone to get what you want.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/voisine/breadwallet/pull/387#issuecomment-233032899>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHw12ScibOlegbTP3Mugp9SM5OzL3lmZks5qV9MzgaJpZM4JM0mQ>
> .
>

ffxsam commented:
I feel bad for @peter-conalgo that this PR turned out to be a bust. @voisine, I honestly don't understand why you have such a strong negative bias against supporting different monetary units in BreadWallet. But, as is your right, it's your project (and it's free) and you're certainly not obligated to listen to the desires/needs of the users. It's just disappointing, frankly.

peter-conalgo commented:
I don't mind pushing this onto the Appstore, but do you have a a guide regarding branding?

I think I'd need to remove the Breadwallet branding, and all access to your servers (in favour of public API's) ... and yet, I don't want to run a wallet or take ownership of the fork's codebase, and I don't have bandwidth to add the other changes that people will ask of me.

I also worry some people will see this as a take over of your fine code, and that is not my intention either. My strong preference is that the PR gets merged.

Would nesting this setting under an "Advanced" label help?

voisine commented:
@peter-conalgo we just ask that you choose a different name and app icon (and other images containing the name) so people aren't confused and mistake it for the official app.

@ffxsam this PR is certainly not a bust. There are several very vocal people who want it.

ffxsam commented:
Sure, the code itself is great. I'm just saying it's a shame that, despite the obvious desire from many users for this feature, it's not going to get included in the app.

voisine commented:
@ffxsam, as explained so many times, there are competing considerations. bitcoin is already complicated enough. A single denomination with two decimal places that everyone standardizes on is our goal, and we're going to do everything we can to help make it happen, including advocating for other bitcoin services and software to make it the default. Adding additional settings/features/configuration to the app is not something we're going to do. There's plenty of other apps, even breadwallet forks, for people who really want that.

ffxsam commented:
If everyone standardizes on mBTC, I could probably get behind that, as it covers most amounts pretty nicely. Based on the current value of XBT (USD$682):

USD$5 = 7.33 mBTC
USD$200 = 293.26 mBTC
USD$50000 = 73313.78 mBTC

@voisine, are you shooting more towards mBTC or bits? I find bits to be kind of ridiculous, akin to buying/selling everything in pennies.

voisine commented:
@peter-conalgo, the breadwallet apis are all public. We just have exchange rate which we are proxying from bitpay at the moment (and the app falls back to using bitpay if ours is down), and then a block explorer for looking up balances on non-wallet private keys, which also falls back to bitpay's public block explorer if ours is down.

There's opt-in, anonymous app usage metrics which you could disable by making BREventManger.shouldAskForPermission always return NO.

Also in BRAPIClient you can disable updateFeatureFlags so you don't get any of the early access stuff like buy bitcoin. These apis are still public though. You're welcome to use them if you want those services in your app.

peter-conalgo commented:
What about the `fee-per-kb` API? Does that also proxy bitpay?

voisine commented:
@peter-conalgo the code will default to the same fee rate that's currently listed on the api. It's only for sudden changes in network conditiond. After bitcoind 0.13 is widely deployed it will use feefilter messages to find the right rate, and won't need that api at all.

voisine commented:
@ffxsam you're not thinking ahead. bitcoin is still tiny, the value of all bitcoin in existence is only as much as a single mid-cap stock, and already a mBTC is worth almost $1. The only reason I'm doing this is because I want to help make bitcoin into a major global currency. The monetary bases of major currencies are measured in the 10's of $trillions. The beauty of bits is that they work even if bitcoin absorbs the value all currencies in existence, they have two decimal places like most existing currencies, and although their value today is inconveniently small, it's still larger than several currencies used by 100's of millions of people every day, like the Indonesian rupiah and Vietnamese dong.

dabura667 commented:
Living in Japan or Korea, paying 100 or 1000 units of currency for a pack of gum is totally reasonable.

We're all happy campers with bits. One of the reasons Japan's most popular wallet is breadwallet.

Future-proofing set aside, America-centric considerations are not the only ones that should be considered imo.

aviadoffer commented:
@debura667 you are happy campers with bits but many others are not. The point is to enable a setting that will make every one happy. If people can choose the right setting from them more people will use breadwallet. No one is against bits or against btc etc. it's about choosing. here in america choosing is part of our freedom :)

tt commented:
What about allowing this to be set through the Settings app? This will allow the flexibility of picking a different base unit while also keeping a clean and uncluttered interface for the common toggles.

It's also what's suggested by [the iOS Human Interface Guidelines](https://developer.apple.com/ios/human-interface-guidelines/interaction/settings/):

> **Expose infrequently changed configuration options in Settings.** The Settings app is a central location for making configuration changes throughout the system, but people must leave your app to get there. It’s far more convenient to adjust settings directly within your app. If you must provide settings that rarely require change, see [Implementing an iOS Settings Bundle](https://developer.apple.com/library/prerelease/content/documentation/Cocoa/Conceptual/UserDefaults/Preferences/Preferences.html) in [Preferences and Settings Programming Guide](https://developer.apple.com/library/prerelease/content/documentation/Cocoa/Conceptual/UserDefaults/Introduction/Introduction.html) for implementation details.

peter-conalgo commented:
I am happy to code the settings-bundle approach, if that approach will be merged by the project owners.

voisine commented:
@tt @peter-conalgo I get that you guys are already used to using full btc, and you don't want to switch. Most other people are not yet using bits as the standard, which only makes it more annoying. I understand that, I really do. This is why I'm willing to keep this PR open on the main project repository for anyone who feels the same way. It's a good PR for this feature, and now it's easier than ever for anyone to have this feature with a few clicks, no coding ability needed.

But please try to understand, our goal is to do anything we can to help bring bitcoin to a much wider audience, and part of that is a *standard* denomination that works for everyone. That denomination is bits, and to make it happen, we have to push for the community to switch standards now while bitcoin is still small. It will only get more difficult the longer we wait. It also means keeping the app as simple as possible, reducing settings and options over time as we find clever ways of making the app cleaner and simpler, not more complex. breadwallet will not have "advanced" settings as long as we can possibly help it. breadwallet is minimalistic. If that's not what you want, there's lots of great geek gadget bitcoin software out there made just for you!

voisine commented:
Just to add one last point, I myself still think in full btc. I find it quite easy to just look at the second comma to the left and imagine it's a decimal point.

samuraisam commented:
Hey @peter-conalgo we've decided to pull this pull request into a branch so there is an official way for people who want to build their own app to include these changes. Could you please update your pull request to point to the `denomination-selector` branch instead of master? We'll try to keep this branch up to date but of course will be open to pull requests from the community to keep it up to date as well.

branch on github: https://github.com/voisine/breadwallet/tree/denomination-selector

peter-conalgo commented:
Ok, I will do that (but I'll need some time for googling the git commands). It's great if you commit to keeping the branch up-to-date, because as it stands nearly any change to the settings menu will break this stuff.

However, I don't see how that helps non-developers. Maybe you should just publicize the existing setting in the app? The slide on navbar thing. It's mostly equivalent and already in the shipping version of the app.

stephancom commented:
Yes, please, merge this to master!  This is my biggest pet peeve about Breadwallet, and has been the main reason I don't use it.  EVERY other place I use bitcoins refers to them in full coins, or at least has an option to switch.  PLEASE PLEASE PLEASE merge this and deploy to the app store.

nshopik commented:
I personally myself prefer bits and will never use it, but I also tend to agree some people may prefer other denominator. Even though I support Aaron that bits should be more universal.

Putting option into settings is good way to avoid cluttering app with settings, which @voisine always mention as argument to avoid adding it. And I don't see this option in any way can't prevent bring bitcoin in wider audience (as Aaron suggest) as long bits are default, and hidden in settings(not in app itself) denominator usage will be minority

jonasschnelli commented:
IMO the denominator depends on the users. Some people may mostly create transactions in the range above 1BTC, some below, etc.

I think a Settings.bundle with a such setting should not hurt and should still conform to @voisine s approach of "UX focused mass market product" (which I kinda think is useful).

aviadoffer commented:
Here is why all those forks just don't work

https://www.reddit.com/r/btc/comments/4w77xq/important_multiple_fake_wallets_making_it_to_the/

This is dangerous, instead of  @voisine stepping up and solving it in the most logical way it creates chaos.

I wish @voisine will step up.

nvk commented:
We still have Coinkite as an App Store vendor, @voisine do you mind if we publish this fork there with a different name to solve this once and for all?

I love this wallet, but "Bits" is fucking driving me insane :)

ffxsam commented:
Those interested may want to also check out Copay which is open source and multi-platform.

bwfan commented:
@voisine  👍 to @nvk offer!

@ffxsam  Copay is so slow, Breadwallet is much better and decentralized.

stephancom commented:
I've started using @bither and I'm liking it.

nvk commented:
Alright folks, we will just call it Breadfork. 

When we find some time we will publish it. I made a very ugly logo to discourage AppStore downloads of the fork. 

![icon180](https://cloud.githubusercontent.com/assets/797193/17531500/1a3fe3be-5e4a-11e6-8c5f-7eebe7e6e299.png)

winneon commented:
I still cannot believe such a small thing like this is not merged into master yet. This whole debacle about not  adding a user-friendly setting over changing a decimal place is so absurd, I cannot believe my eyes. I understand @voisine's "vision" about keeping this app's default setting to bits, but it's anti-consumer and anti-user-friendly to not add something as obvious as a setting to change a decimal place. Some people make transactions larger than 1BTC regularly, and seeing a large number can be annoying. BTC is also the standard for pretty much every service using Bitcoin in the first place.

@voisine, I do hope you realize that having such a petty debate about something that shouldn't even be an argument is causing you to lose reputation and users of your app.

dabura667 commented:
@winneon You can change the unit in current breadwallet from the app store.

When selecting fiat currency display swipe the bitcoin to fiat conversion rate text area left or right.

voisine commented:
@winneon as I've mentioned, I want to help change the fact that BTC is the standard unit. This requires annoyance and irritation in the short term, in service of a long term goal. I'm happy to sacrifice my reputation in the service of expanding bitcoin adoption to the rest of the world, if it helps.

winneon commented:
> You can change the unit in current breadwallet from the app store.
> 
> When selecting fiat currency display swipe the bitcoin to fiat conversion rate text area left or right.

Wait, so if this feature is already implemented, albeit in a somewhat roundabout way, why the hell is this large debacle over absolutely nothing? I don't understand why the developers are making a stance towards keeping bits as the only options when they patched in a way to change it. It's contradictory.

erikbos commented:
+1 for including the capable to change the denomination!

nvk commented:
Thx for adding this on the new version! 🎆
